// @stdlib/_tools/remark/plugins/remark-lint-html-section-structure/lib/main.js
/**
* @license Apache-2.0
*
* Copyright (c) 2025 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var logger = require( 'debug' );
var rule = require( 'unified-lint-rule' );
var visit = require( 'unist-util-visit' );


// VARIABLES //

var debug = logger( 'remark-lint-html-section-structure' );
var SECTION_START = /<section(?:\s+class="([^"]*)")?>/;
var SECTION_END = /<\/section>$/;
var SECTION_END_WITH_COMMENT = /<\/section>\s*<!--\s*\/\.([^*]*)\s*-->/;
var END_COMMENT = /^\s*<!--\s*\/\.([^\s]*)\s*-->/;
var END_COMMENT_SIMPLE = /<!--\s*\/\.([^>]+)-->/;
var htmlSectionStructureRule;


// MAIN //

/**
* Validates HTML section structure in README files.
*
* @private
* @param {Node} tree - abstract syntax tree (AST)
* @param {File} file - virtual file
* @param {Object} options - options
* @param {Callback} clbk - callback to invoke upon completion
* @returns {void}
*/
function linter( tree, file, options, clbk ) {
	var isValidWhitespace;
	var combinedMatch;
	var rootSections;
	var sectionStack;
	var commentClass;
	var sectionMatch;
	var simpleMatch;
	var commentNode;
	var commentLine;
	var structure;
	var positions;
	var className;
	var nextNode;
	var endMatch;
	var endLine;
	var lineGap;
	var current;
	var match;
	var value;
	var data;
	var msg;
	var pos;
	var i;

	debug( 'Linting file: %s', file.path || '' );

	sectionStack = [];
	rootSections = [];
	structure = [];
	positions = {};

	visit( tree, 'html', visitor );

	debug( 'Final section stack length: %d', sectionStack.length );

	if ( sectionStack.length > 0 ) {
		debug( 'Found unclosed sections: %d', sectionStack.length );
		for ( i = 0; i < sectionStack.length; i++ ) {
			className = sectionStack[ i ].className || '[no class specified]';
			pos = sectionStack[ i ].position.start.line + ':' + sectionStack[ i ].position.start.column;
			msg = pos + '   error   Unclosed section tag. Missing </section> closing tag for section with class="' + className + '". Add a closing tag to fix.   unclosed-section';
			debug( msg );
			file.message( msg, sectionStack[ i ] );
		}
	}

	debug( 'Finished linting: %s', file.path || '' );
	return clbk();

	/**
	* Callback invoked upon finding a matching node.
	*
	* @private
	* @param {Object} node - AST node
	* @param {number} index - position of `node` in `parent`
	* @param {Object} parent - parent AST node
	* @returns {void}
	*/
	function visitor( node, index, parent ) {
		// Store position for tracking purposes
		if ( node.position ) {
			positions[ node.position.start.line ] = node;
		}

		// Check if this is a combined section end with comment on the same line (which is an error)
		combinedMatch = SECTION_END_WITH_COMMENT.exec( node.value );
		if ( combinedMatch ) {
			debug( 'Found combined section end with comment: %s', node.value );
			pos = node.position.start.line + ':' + node.position.start.column;
			msg = pos + '   error   Missing empty line between closing section tag and comment. There should be an empty line between </section> and <!-- /.' + combinedMatch[1] + ' -->   missing-empty-line';
			debug( msg );
			file.message( msg, node );

			// Make sure to pop the section stack for this case - otherwise we'll get "unclosed section" errors
			if ( sectionStack.length > 0 ) {
				debug( 'Popping section stack for combined section end+comment' );
				sectionStack.pop();
			}
			return;
		}

		// Check if this is a section start tag
		sectionMatch = SECTION_START.exec( node.value );
		if ( sectionMatch ) {
			debug( 'Found a section start tag with class: %s', sectionMatch[1] || '[none]' );

			data = {
				'node': node,
				'className': sectionMatch[1] || '',
				'position': node.position,
				'children': [],
				'index': index,
				'parent': parent
			};

			if ( sectionStack.length === 0 ) {
				// This is a root level section
				rootSections.push( data );
				structure.push( data );
			} else {
				// This is a nested section, add to parent's children
				sectionStack[sectionStack.length - 1].children.push( data );
			}

			sectionStack.push( data );
			return;
		}

		// Check if this is a section end tag
		endMatch = SECTION_END.exec( node.value );
		if ( endMatch ) {
			debug( 'Found a section end tag' );

			if ( sectionStack.length === 0 ) {
				debug( 'Orphaned section end tag found' );
				pos = node.position.start.line + ':' + node.position.start.column;
				msg = pos + '   error   Orphaned section closing tag. Found </section> without a matching opening <section> tag. Remove this tag or add a corresponding opening tag.   orphaned-section-end';
				debug( msg );
				file.message( msg, node );
				return;
			}

			// Special handling for direct closing tag followed by a comment (no text node between)
			if ( index + 1 < parent.children.length ) {
				nextNode = parent.children[index + 1];

				// Look at the next node to check if it's a comment
				if ( nextNode.type === 'html' ) {
					// Check if the next node is a comment but NOT on the same line
					// This avoids double-reporting with the SECTION_END_WITH_COMMENT check above
					match = END_COMMENT.exec( nextNode.value );
					simpleMatch = END_COMMENT_SIMPLE.exec( nextNode.value );

					// If either regex matches, we've found a comment
					if ( match || simpleMatch ) {
						// Check if there are at least 2 line breaks between them
						endLine = node.position.end.line;
						commentLine = nextNode.position.start.line;
						lineGap = commentLine - endLine;

						debug( 'Gap between section end and comment: %d lines', lineGap );

						// We require at least 2 lines of gap for an empty line
						if ( lineGap < 2 ) {
							pos = node.position.start.line + ':' + node.position.start.column;
							msg = pos + '   error   Missing proper empty line after closing section tag. There should be an empty line between </section> and the closing comment.   missing-empty-line-after-section';
							debug( msg );
							file.message( msg, node );
						}

						// Verify the comment matches the class name:
						current = sectionStack[sectionStack.length - 1];

						// Get class from whichever regex matched:
						if ( match ) {
							commentClass = match[1];
						} else {
							commentClass = simpleMatch[1].trim();
						}

						// Verify the comment matches the class name:
						if ( current.className && commentClass !== current.className ) {
							debug( 'Mismatched section class in comment: expected %s, got %s', current.className, commentClass );
							pos = nextNode.position.start.line + ':' + nextNode.position.start.column;
							msg = pos + '   error   Mismatched section class in closing comment. Opening tag has class="' + current.className + '" but closing comment is <!-- /.' + commentClass + ' -->. They should match.   mismatched-section-class';
							debug( msg );
							file.message( msg, nextNode );
						}

						// Pop the section stack before returning
						sectionStack.pop();
						return;
					}
				}
			}

			// Standard handling for other cases with whitespace between section end and comment...
			if ( index + 2 < parent.children.length ) {
				nextNode = parent.children[index + 1];
				commentNode = parent.children[index + 2];

				// Check if the next node is a proper whitespace with empty line:
				isValidWhitespace = false;
				if (nextNode.type === 'text') {
					value = nextNode.value;

					// Look for at least two consecutive newlines (representing an empty line):
					isValidWhitespace = ((/(\r?\n)\s*(\r?\n)/).test(value));
				}

				if ( !isValidWhitespace ) {
					debug( 'Invalid whitespace after section end' );
					pos = node.position.start.line + ':' + node.position.start.column;
					msg = pos + '   error   Missing proper empty line after closing section tag. There should be an empty line between </section> and the closing comment.   missing-empty-line-after-section';
					debug( msg );
					file.message( msg, node );
				}

				// Check if the following node is the expected comment
				if ( commentNode.type === 'html' ) {
					match = END_COMMENT.exec( commentNode.value );
					if ( match ) {
						current = sectionStack[sectionStack.length - 1];

						// Verify the comment matches the class name
						if ( current.className && match[1] !== current.className ) {
							debug( 'Mismatched section class in comment: expected %s, got %s', current.className, match[1] );
							pos = commentNode.position.start.line + ':' + commentNode.position.start.column;
							msg = pos + '   error   Mismatched section class in closing comment. Opening tag has class="' + current.className + '" but closing comment is <!-- /.' + match[1] + ' -->. They should match.   mismatched-section-class';
							debug( msg );
							file.message( msg, commentNode );
						}
					} else {
						// The HTML node exists but it's not a proper closing comment
						debug( 'Invalid closing comment format' );
						pos = commentNode.position.start.line + ':' + commentNode.position.start.column;
						msg = pos + '   error   Invalid section closing comment format. Expected <!-- /.' + sectionStack[sectionStack.length - 1].className + ' -->   invalid-closing-comment';
						debug( msg );
						file.message( msg, commentNode );
					}
				} else {
					// The node after whitespace is not an HTML node (comment)
					debug( 'Node after whitespace is not an HTML comment' );
					pos = node.position.start.line + ':' + node.position.start.column;
					msg = pos + '   error   Missing section closing comment. Section closing tag should be followed by <!-- /.' + sectionStack[sectionStack.length - 1].className + ' --> comment.   missing-section-comment';
					debug( msg );
					file.message( msg, node );
				}
			} else {
				// There's not enough nodes after the section closing tag
				debug( 'Not enough nodes after section end' );
				pos = node.position.start.line + ':' + node.position.start.column;
				msg = pos + '   error   Missing section closing comment. Section closing tag should be followed by <!-- /.' + sectionStack[sectionStack.length - 1].className + ' --> comment on a separate line.   missing-section-comment';
				debug( msg );
				file.message( msg, node );
			}

			// Pop the section stack
			sectionStack.pop();
		}
	}
}

htmlSectionStructureRule = rule( 'remark-lint:html-section-structure', linter );


// EXPORTS //

module.exports = htmlSectionStructureRule;
