// @stdlib/_tools/eslint/rules/tsdoc-declarations-doctest/lib/main.js
/**
* @license Apache-2.0
*
* Copyright (c) 2025 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

/* eslint-disable max-len, max-lines */

'use strict';

// MODULES //

var vm = require( 'vm' );
var resolve = require( 'path' ).resolve;
var logger = require( 'debug' );
var resolveParentPath = require( '@stdlib/fs/resolve-parent-path' ).sync;
var readJSON = require( '@stdlib/fs/read-json' ).sync;
var dirname = require( '@stdlib/utils/dirname' );
var hasOwnProp = require( '@stdlib/assert/has-own-property' );
var contains = require( '@stdlib/assert/contains' );
var isNull = require( '@stdlib/assert/is-null' );
var isNumber = require( '@stdlib/assert/is-number' );
var replace = require( '@stdlib/string/replace' );
var endsWith = require( '@stdlib/string/ends-with' );
var trim = require( '@stdlib/string/trim' );
var reEOL = require( '@stdlib/regexp/eol' );
var escapeRegExpString = require( '@stdlib/utils/escape-regexp-string' );
var objectKeys = require( '@stdlib/utils/keys' );
var tryRequire = require( '@stdlib/utils/try-require' );
var compareValues = require( '@stdlib/_tools/doctest/compare-values' );
var createAnnotationValue = require( '@stdlib/_tools/doctest/create-annotation-value' );
var createVMScope = require( './create_vm_scope.js' );
var addPackageToScope = require( './add_package_to_scope.js' );
var SCOPE_DEFAULTS = require( './scope_defaults.json' );


// VARIABLES //

var debug = logger( 'tsdoc-doctest' );
var RE_TSDOC = /\/\*\*[\s\S]+?\*\//g;
var RE_EXAMPLE = /@example\s*([\s\S]*?)(?=\n\s*@\w|\*\/|$)/g;
var RE_ANNOTATION = /(?:\n|^)(?:var|let|const)? ?([a-zA-Z0-9._]+) ?=?[^;]*;\n\/\/ ?(returns|([A-Za-z][A-Za-z_0-9]*)? ?=>|throws) ?([\s\S]*?)(?=\n|$)/g;
var RE_COMMENT_PREFIX = /^\s*\*\s?/gm;
var DEFAULT_ERROR_LOC = {
	'start': {
		'line': 1,
		'column': 0
	},
	'end': {
		'line': 1,
		'column': 0
	}
};
var regexCache = {};
var lineCountCache = {};
var rule;


// FUNCTIONS //

/**
* Gets or creates a cached regex for finding return annotations.
*
* @private
* @param {string} annotationType - type of annotation (returns, throws, etc.)
* @param {string} value - expected value
* @returns {RegExp} compiled regular expression
*/
function getAnnotationRegex( annotationType, value ) {
	var escapedValue;
	var pattern;
	var key;

	escapedValue = escapeRegExpString( value );
	key = annotationType + '::' + escapedValue;

	if ( !regexCache[ key ] ) {
		pattern = '// ' + annotationType + '\\s+' + escapedValue;
		regexCache[ key ] = new RegExp( pattern );
	}
	return regexCache[ key ];
}

/**
* Counts the number of lines in the given string.
*
* @private
* @param {string} str - input string
* @returns {number} number of lines
*/
function countLines( str ) {
	var lines;

	if ( hasOwnProp( lineCountCache, str ) ) {
		return lineCountCache[ str ];
	}
	lines = str.split( reEOL.REGEXP );
	lineCountCache[ str ] = lines.length - 1;
	return lineCountCache[ str ];
}

/**
* Searches for a variable in scope matching the expected value.
*
* @private
* @param {Object} scope - VM scope
* @param {*} expected - expected value to search for
* @returns {string} variable name or `?` if no match found
*/
function findName( scope, expected ) {
	var keys = objectKeys( scope );
	var key;
	var i;
	for ( i = 0; i < keys.length; i++ ) {
		key = keys[ i ];
		if (
			!contains( SCOPE_DEFAULTS, key ) &&
			compareValues( scope[ key ], expected ) === null
		) {
			return key;
		}
	}
	return '?';
}

/**
* Resolves the implementation path from the nearest parent `package.json`.
*
* @private
* @param {string} filepath - TypeScript declaration file path
* @returns {(string|null)} resolved implementation path or null if not found
*/
function resolveImplementationPath( filepath ) {
	var pkgJSON;
	var pkgPath;
	var opts;

	// Find the nearest package.json:
	opts = {
		'dir': dirname( resolve( filepath ) )
	};
	pkgPath = resolveParentPath( 'package.json', opts );
	if ( !pkgPath ) {
		debug( 'Could not find package.json for: ' + filepath );
		return null;
	}

	// Read and parse package.json:
	pkgJSON = readJSON( pkgPath );
	if ( pkgJSON instanceof Error ) {
		debug( 'Could not read/parse package.json: ' + pkgPath + '. Error: ' + pkgJSON.message );
		return null;
	}

	// Get the main entry point:
	return resolve( dirname( pkgPath ), pkgJSON.main || './lib' );
}

/**
* Cleans TSDoc comment by removing comment prefixes.
*
* @private
* @param {string} comment - TSDoc comment
* @returns {string} cleaned comment
*/
function cleanTSDocComment( comment ) {
	// Remove opening /** and closing */
	var cleaned = replace( replace( comment, /^\/\*\*/, '' ), /\*\/$/, '' );

	// Remove * at the beginning of lines
	return replace( cleaned, RE_COMMENT_PREFIX, '' );
}

/**
* Processes a single example code block, finding and validating annotations.
*
* @private
* @param {string} code - example code
* @param {number} commentIdx - comment index
* @param {Array} comments - all comments
* @param {Object} scope - VM execution scope
* @param {Function} report - error reporting function
* @param {Object} opts - options object
* @param {Object} sourceCode - source code object
*/
function processExampleCode( code, commentIdx, comments, scope, report, opts, sourceCode ) {
	var returnAnnotationPattern;
	var valueStartInAnnotation;
	var commentStartIdx;
	var annotationMatch;
	var annotationStart;
	var exampleStartIdx;
	var annotationEnd;
	var intermediary;
	var replacement;
	var exampleIdx;
	var remaining;
	var expected;
	var codeIdx;
	var actual;
	var last;
	var line;
	var type;
	var out;
	var loc;
	var msg;
	var arr;

	last = 0;
	RE_ANNOTATION.lastIndex = 0;
	try {
		arr = RE_ANNOTATION.exec( code );
		while ( !isNull( arr ) ) {
			intermediary = code.substring( last, arr.index );
			last = arr.index + arr[ 0 ].length;
			if ( intermediary ) {
				vm.runInContext( intermediary, scope );
			}

			// Calculate line of current code chunk within the comment:
			commentStartIdx = sourceCode.text.indexOf( comments[ commentIdx ] );

			// Find the annotation in the original comment:
			returnAnnotationPattern = getAnnotationRegex( arr[ 2 ], arr[ 4 ] );
			annotationMatch = comments[ commentIdx ].match( returnAnnotationPattern );

			if ( !annotationMatch ) {
				continue;
			}

			codeIdx = comments[ commentIdx ].indexOf( annotationMatch[ 0 ] );
			line = countLines( sourceCode.text.substring( 0, commentStartIdx + codeIdx ) ) + 1;
			loc = {
				'start': {
					'line': line,
					'column': 0
				},
				'end': {
					'line': line + countLines( arr[ 0 ] ),
					'column': arr[ 0 ].length
				}
			};

			// Run code preceding return annotation and capture result:
			try {
				out = vm.runInContext( arr[ 0 ], scope );

				// Determine actual value based on annotation form:
				if ( arr[ 3 ] ) {
					// Expression form with named variable: someExpression; // variable => value
					actual = vm.runInContext( arr[ 3 ], scope );
				} else if ( arr[ 1 ] && arr[ 1 ] !== 'console.log' && !arr[ 2 ].match( /^=>/ ) ) {
					// Assignment form: var x = expression; // returns value
					actual = scope[ arr[ 1 ] ];
				} else {
					// Bare expression form (incl. console.log): someExpression; // => value
					actual = out;
				}
				expected = arr[ 4 ];
				msg = compareValues( actual, expected );
				if ( msg ) {
					opts.includeDecimal = isNumber( actual ) && contains( expected, '.' );
					replacement = createAnnotationValue( actual, opts );

					// Find the position of the return value in the annotation:
					valueStartInAnnotation = annotationMatch[ 0 ].indexOf( arr[ 4 ] );
					loc.range = [
						// Position of arr[4] start in source text:
						commentStartIdx + codeIdx + valueStartInAnnotation,

						// Position of arr[4] end in source text:
						commentStartIdx + codeIdx + valueStartInAnnotation + arr[ 4 ].length
					];
					report( loc, msg, replacement );
				}
			} catch ( err ) {
				type = '<'+err.name+'>';
				if ( arr[ 2 ] !== 'throws' ) {
					if ( arr[ 3 ] ) {
						msg = 'Encountered an error: `'+err.message+'`.';
					} else {
						msg = 'Encountered an error while running code: `'+err.message+'`. ';
						msg += 'Did you mean to include a `// throws '+type+'` annotation instead of `// '+arr[ 2 ]+' '+arr[ 4 ]+'`?';
					}
				}
				else if ( arr[ 4 ] !== type ) {
					msg = 'Code should have a `// throws '+type+'` annotation, but received: `// '+arr[ 2 ]+' '+arr[ 4 ]+'`';
				}
				if ( msg ) {
					replacement = ( arr[ 3 ] ) ? findName( scope, arr[ 4 ] ) + ' => ' + arr[ 4 ] : 'throws '+type;

					// Find annotation part in the match (after "// "):
					annotationStart = arr[ 0 ].indexOf( '// ' ) + 3;
					annotationEnd = arr[ 0 ].length;
					loc.range = [
						commentStartIdx + codeIdx + annotationStart,
						commentStartIdx + codeIdx + annotationEnd
					];
					report( loc, msg, replacement );
				}
			}
			arr = RE_ANNOTATION.exec( code );
		}

		remaining = code.substring( last );
		if ( remaining ) {
			vm.runInContext( remaining, scope );
		}
	} catch ( err ) {
		// Calculate the line number for the example that caused the error...
		exampleStartIdx = sourceCode.text.indexOf( comments[ commentIdx ] );
		exampleIdx = comments[ commentIdx ].indexOf( code );
		line = countLines( sourceCode.text.substring( 0, exampleStartIdx + exampleIdx ) ) + 1;
		loc = {
			'start': {
				'line': line,
				'column': 0
			},
			'end': {
				'line': line + countLines( code ),
				'column': 0
			}
		};

		// Always report errors so programmer errors are not swallowed:
		report( loc, 'Encountered an error while running example code: '+err.message );
	}
}

/**
* Rule for validating that return annotations in TypeScript declaration examples match the actual output.
*
* @param {Object} context - ESLint context
* @returns {Object} validators
*/
function main( context ) {
	var sourceCode;
	var filename;
	var dir;

	sourceCode = context.sourceCode;
	filename = context.filename;

	// Only process TypeScript declaration files:
	if ( !endsWith( filename, '.d.ts' ) ) {
		return {};
	}

	dir = dirname( filename );

	/**
	* Reports the error message.
	*
	* @private
	* @param {Object} loc - error location info
	* @param {string} msg - error message
	* @param {string} replacement - fixed return annotation
	*/
	function report( loc, msg, replacement ) {
		var result = {
			'message': msg,
			'loc': loc
		};
		if ( replacement ) {
			result.fix = fix;
		}
		context.report( result );

		/**
		* Applies a fix to the offending code.
		*
		* @private
		* @param {Object} fixer - object to apply a fix
		* @returns {Object} fixing object
		*/
		function fix( fixer ) {
			return fixer.replaceTextRange( loc.range, replacement );
		}
	}

	/**
	* Validates examples in TSDoc comments.
	*
	* @private
	*/
	function validate() {
		var implPath;
		var comments;
		var examples;
		var cleaned;
		var example;
		var comment;
		var scope;
		var opts;
		var code;
		var pkg;
		var i;
		var j;

		// Clear caches to prevent memory leaks:
		regexCache = {};
		lineCountCache = {};

		// Resolve implementation path from nearest package.json:
		implPath = resolveImplementationPath( filename );
		if ( !implPath ) {
			debug( 'Could not resolve implementation path from package.json for: ' + filename );
			report( DEFAULT_ERROR_LOC, 'Could not resolve implementation path from package.json for this declaration file.' );
			return;
		}

		// Try to load the implementation directly:
		pkg = tryRequire( implPath );
		if ( pkg instanceof Error ) {
			debug( 'Could not load implementation: ' + implPath + '. Error: ' + pkg.message );
			report( DEFAULT_ERROR_LOC, 'Could not load implementation: '+implPath+' ('+pkg.message+').' );
			return;
		}

		// Get all TSDoc comments:
		comments = sourceCode.text.match( RE_TSDOC );
		if ( !comments ) {
			return;
		}

		// Create a single VM scope and context for all examples:
		scope = createVMScope( dir, filename );
		addPackageToScope( scope, pkg, sourceCode.text );
		vm.createContext( scope );

		opts = {
			'includeDecimal': false
		};

		for ( i = 0; i < comments.length; i++ ) {
			comment = comments[ i ];
			cleaned = cleanTSDocComment( comment );

			// Extract @example blocks:
			RE_EXAMPLE.lastIndex = 0;
			examples = [];
			example = RE_EXAMPLE.exec( cleaned );
			while ( example !== null ) {
				examples.push( example[ 1 ] );
				example = RE_EXAMPLE.exec( cleaned );
			}

			for ( j = 0; j < examples.length; j++ ) {
				code = trim( examples[ j ] );

				// Process the example code and validate annotations (reusing VM context):
				processExampleCode( code, i, comments, scope, report, opts, sourceCode );
			}
		}
	}

	return {
		'Program': validate
	};
}


// MAIN //

rule = {
	'meta': {
		'type': 'problem',
		'docs': {
			'description': 'ensure return annotations in TSDoc examples match the actual output'
		},
		'fixable': 'code',
		'schema': []
	},
	'create': main
};


// EXPORTS //

module.exports = rule;
